components/learn/modules/ROOT/pages/unit-testing.adoc

```diff
@@ -0,0 +1,174 @@
+
+What we will use is called a _local blockchain_: a slimmed down version of the real thing, disconnected from the Internet, running on your machine. This will simplify things quite a bit: you won't need to go get Ether, and you won't have to wait for blocks to be mined.
+
+To aid us in this we'll use the https://github.com/OpenZeppelin/openzeppelin-test-environment#readme[*OpenZeppelin Test Environment*], a JavaScript library that will take care of setting up our local blockchain.
+
+TIP: If you've read the xref:interact.adoc['Interacting with Your Contracts' guide], you will already be familiar with https://github.com/trufflesuite/ganache-cli/[Ganache]. Test Environment works in a similar way, but is much easier to use.
+
+To install the OpenZeppelin Test Environment, run:
+
+```bash
+$ npm install --save-dev @openzeppelin/test-enviroment
+```
+
+Once you `require` the library from your JavaScript code, it will automatically run a local testing blockchain for you. It also exports variables and functions related to this environment, such as a list of `accounts` that have been pre-funded with Ether, and convenient ways to load your contracts from their compiled artifacts.
+
+```javascript
+const { accounts, contract } = require('@openzeppelin/test-environment#readme');
+
+// Use the different accounts, which are unlocked and funded with Ether
+const [ admin, deployer, user ] = accounts;
+
+// Create a contract object from a compilation artifact
+const MyContract = contract.fromArtifact('MyContract');
+```
+
+For detailed information on these exported values and their usage, refer to Test Environment's https://github.com/OpenZeppelin/openzeppelin-test-environment/blob/master/docs/modules/ROOT/pages/api.adoc[API reference].
+
+[[writing-unit-tests]]
+== Writing Unit Tests
+
+In order to actually run your tests, you will need to also install a JavaScript _test runner_. You are free to use any of the https://github.com/OpenZeppelin/openzeppelin-test-environment/blob/master/docs/modules/ROOT/pages/setup.adoc#test-runners[recommended ones]: for this guide, we'll pick https://mochajs.org/[Mocha] with https://www.chaijs.com/[Chai] assertions.
+
+```bash
+$ npm install --save-dev mocha chai
+```
+
+Create a `test` directory: this is where you will keep your test files. These are best structured by mirroring xref:write-contracts#setting-up-a-solidity-project[the `contracts` directory]: for each `.sol` file there, create a corresponding `.test.js` file.
+
+Time to write our first tests! These will test properties of the `SetAndGet` contract xref:write-contracts#set-and-get-contract[from previous guides].
```

Comment:
Feedback I've got from the book is that it's not always clear in which file we should place the code. Even though you have explained it in the previous paragraph, it may be simpler for the user if you either explicitly name the file in this paragraph, or add it as a leading comment in the code block, like `// test/contracts/SetAndGet.sol`.

Reply:
Ohh, I like the leading comment bit